### PR TITLE
v2: Add support for the distribution of kABI-related files inside of the kernel tarball

### PR DESCRIPTION
**Pull request by ckalina-rh**
Introduce the support for appending to the generated tarball.

Changes in v2:
 - fixed Travis/flake8 issues with 80 line intend and whitespaces around `[` `]`, no more `flake8` errors encountered;
```
./skt/kernelbuilder.py:350:80: E501 line too long (80 > 79 characters)
./skt/kernelbuilder.py:351:80: E501 line too long (80 > 79 characters)
./skt/kernelbuilder.py:425:28: E201 whitespace after '['
./skt/kernelbuilder.py:425:50: E202 whitespace before ']'
./skt/kernelbuilder.py:449:26: E201 whitespace after '['
./skt/kernelbuilder.py:449:46: E202 whitespace before ']'
```
```
$ flake8 --show-source ./kernelbuilder.py
$
```

### CODE REVIEW

File: skt/kernelbuilder.py [hunk 1]
```diff
@@ -327,13 +336,53 @@ def make_rpm_repo(self, rpm_files):
 
         return repo_dir
 
+    def find_extras(self, rel_path, callback=None, followlinks=True):
```
Comment (veruu) [inline skt/kernelbuilder.py]: Given the docstring about `followlinks` says "Set this to True only if you're sure that this won't cause loops", I'm not sure default value of `True` is a good idea. Could you use `False` there so there are no unexpected problems? It would be also easier to reuse the method for other extras that way too.

File: skt/kernelbuilder.py [hunk 2]
```diff
@@ -327,13 +336,53 @@ def make_rpm_repo(self, rpm_files):
...
+                         that support them. Set this to True only if you're
+                         sure that this won't cause loops.
+
+        Returns:
+            A (possibly pruned) list of paths recursively found in
+            `rel_path`. Returns an empty list if no matches were found
+            or an exception occured during pruning.
+        """
+        source_dir_abs = os.path.abspath(self.source_dir)
+        target_dir = source_dir_abs + rel_path
```
Comment (veruu) [inline skt/kernelbuilder.py]: Could you please use `os.path.join` or `skt.misc.join_with_slash` for path joins?

I also see that when you call the method you use a starting slash with the directory path which makes the path look like absolute path which is confusing. Using one of the mentioned functions would help with that too.

File: skt/kernelbuilder.py [hunk 3]
```diff
@@ -327,13 +336,53 @@ def make_rpm_repo(self, rpm_files):
...
+
     def find_tarball(self):
         """
         Find a tarball in the buildlog.
 
         Returns:
-            The full path to the tarball (as a string), or None if the tarball
-            line was not found in the log.
+            The full path to the tarball (as a string), or None if the
+            tarball line was not found in the log.
```
Comment (veruu) [inline skt/kernelbuilder.py]: This change looks unrelated to the patches, could you leave it out?

File: skt/kernelbuilder.py [hunk 4]
```diff
@@ -357,6 +406,49 @@ def find_tarball(self):
...
+
+        Args:
+            tarball: Path to a tarball.
+            extras:  A list of file paths (relative to the kernel tree root) to
+                     append.
+        """
+        if not extras:
+            return
+
+        tarball_tmp = os.path.abspath(tarball)
```
Comment (veruu) [inline skt/kernelbuilder.py]: Tarball path should always be absolute at this point so this step is not necessary.

File: skt/kernelbuilder.py [hunk 5]
```diff
@@ -327,13 +336,53 @@ def make_rpm_repo(self, rpm_files):
...
+        """
+        source_dir_abs = os.path.abspath(self.source_dir)
+        target_dir = source_dir_abs + rel_path
+
+        result = []
+
+        logging.info("looking for extras in: %s", target_dir)
+        for root, _, files in os.walk(target_dir, followlinks=followlinks):
+            result += [
+                os.path.relpath(root + '/' + f, source_dir_abs)
```
Comment (veruu) [inline skt/kernelbuilder.py]: Same with the path joining here, please use `os.path.join` or `skt.misc.join_with_slash`

File: skt/kernelbuilder.py [hunk 6]
```diff
@@ -387,6 +479,10 @@ def compile_kernel(self, timeout=60 * 60 * 12):
             + kernel_build_argv
         )
 
+        # Generate kABI whitelist information for RHEL kernels.
+        if self.cfgtype == 'rh-kernels':
```
Comment (veruu) [inline skt/kernelbuilder.py]: `cfgtype` should be `rh-configs`, otherwise this will never run.

**Comment by veruu** [comment]
Please feel free to integrate the changes and force-push so this pull request gets updated and we don't loose the discussion points, instead of opening a new one. Thanks!

**Comment by ckalina-rh** [comment]
> Please feel free to integrate the changes and force-push so this pull request gets updated and we don't loose the discussion points, instead of opening a new one. Thanks!

Done, thanks for the review!

**Comment by veruu** [comment]
Looking ok on the code side. @cki-bot can you please test this?

**Comment by ckalina-rh** [comment]
@veruu added some logging as well as return code checks for g(un)zip, make and cpio

**Comment by veruu** [comment]
Yes @cki-bot we know it failed, and already have a new version here. Can you test it again?

**Comment by ckalina-rh** [comment]
@cki-bot humbly asking to test again, please

**Comment by ckalina-rh** [comment]
@cki-bot test, please

**Comment by ckalina-rh** [comment]
@cki-bot test, please

**Comment by ckalina-rh** [comment]
@cki-bot test, please

**Comment by ckalina-rh** [comment]
@cki-bot test, please

**Comment by ckalina-rh** [comment]
older cpio versions do not support -D; remedy that by passing cwd to Popen instead
@cki-bot test, please

**Comment by ckalina-rh** [comment]
@cki-bot test

**Comment by ckalina-rh** [comment]
Due to possible future directory layout change, specify another kABI directory. For details, see https://github.com/CKI-project/skt/compare/12abab33d3cac11e5c4ea0b3c6095bd98ffd70d1..073350574e09a6f64338c6ea7719f3d861427635

**Comment by veruu** [comment]
(sorry from the bot spam to everyone who has email notifications enabled, we are implementing more security measures and it grew a consciousness of its own overnight. everything should be cleaned up now)

That said, @ckalina-rh can we close this pull? The changes are now implemented directly in the pipeline and skt is not used to build the kernels anymore.

**Comment by ckalina-rh** [comment]
Sure, closing.
